### PR TITLE
Support mysql `partition` to table selection

### PR DESCRIPTION
**Pull request by chunshao90**
Support mysql partition selection, example:
```
SELECT * FROM employees PARTITION (p0, p2)
```
Refer to https://dev.mysql.com/doc/refman/8.0/en/partitioning-selection.html

### CODE REVIEW

**Comment by chunshao90** [comment]
@alamb PTAL

**Comment by alamb** [comment]
Sorry for the delay -- I was away last week. I am looking at this PR now

File: src/ast/query.rs [hunk 1]
```diff
@@ -665,6 +665,8 @@ pub enum TableFactor {
         /// Optional version qualifier to facilitate table time-travel, as
         /// supported by BigQuery and MSSQL.
         version: Option<TableVersion>,
+        /// Partition selection, supported by MySQL.
```
Comment (alamb) [inline src/ast/query.rs:668]:
```suggestion
        /// [Partition selection](https://dev.mysql.com/doc/refman/8.0/en/partitioning-selection.html), supported by MySQL.
```

File: src/parser/mod.rs [hunk 2]
```diff
@@ -6210,6 +6210,20 @@ impl<'a> Parser<'a> {
         } else {
             let name = self.parse_object_name()?;
 
+            let partitions = if dialect_of!(self is MySqlDialect)
```
Comment (alamb) [inline src/parser/mod.rs:6213]: Generally this crate also allows all syntax for the `GenericDialect`, such as

https://github.com/sqlparser-rs/sqlparser-rs/blob/4c3a4ad5a87ef8b4dd24a6e8540243e14fbfa148/src/parser/mod.rs#L756

File: src/parser/mod.rs [hunk 3]
```diff
@@ -6210,6 +6210,20 @@ impl<'a> Parser<'a> {
...
+                && self.parse_keyword(Keyword::PARTITION)
+            {
+                let mut partitions = self.parse_comma_separated(|p| p.parse_tuple(true, false))?;
```
Comment (alamb) [inline src/parser/mod.rs:6216]: As written I think this only checks the first partition for having more than one tuple element (what about values after index 0?)

I think if you used `parse_expr` here this code would be significantly simpler and there would be no need for error checking

```suggestion
                let mut partitions = self.parse_comma_separated(Parser::parse_expr)?;
```

File: src/ast/query.rs [hunk 4]
```diff
@@ -665,6 +665,8 @@ pub enum TableFactor {
...
+        partitions: Vec<Expr>,
```
Comment (alamb) [inline src/ast/query.rs:669]: I wonder why this is an `Expr` ? It seems like each element is a `partition_name` 

https://dev.mysql.com/doc/refman/8.0/en/create-table.html

So perhaps this should be something like 

```suggestion
        partitions: Vec<Ident>,
```

And then call https://docs.rs/sqlparser/latest/sqlparser/parser/struct.Parser.html#method.parse_identifier to parse them below?

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @chunshao90  -- this is looking  pretty close. I left some suggestions for your consideration

File: src/parser/mod.rs (hunk 3 above)
Comment (chunshao90) [reply src/parser/mod.rs]: Returns an error if there are multiple tuples.

File: src/ast/query.rs (hunk 4 above)
Comment (chunshao90) [reply src/ast/query.rs]: done

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @chunshao90

**Comment by alamb** [comment]
I'll clean up the clippy failures

**Comment by alamb** [comment]
Thanks again @chunshao90 !

**Pull request merged**